server.go

```diff
@@ -1825,10 +1826,25 @@ func (s *server) handleRelayInvMsg(state *peerState, msg relayMsg) {
 			return
 		}
 
-		// If the inventory is a block and the peer prefers headers,
-		// generate and send a headers message instead of an inventory
-		// message.
-		if msg.invVect.Type == wire.InvTypeBlock && sp.WantsHeaders() {
+		// Ignore peers that do not have the required service flags.
+		if !hasServices(sp.Services(), msg.reqServices) {
+			return
+		}
+
+		// Filter duplicate block announcements.
+		iv := msg.invVect
+		isBlockAnnouncement := iv.Type == wire.InvTypeBlock
+		if isBlockAnnouncement {
+			if sp.announcedBlock != nil && *sp.announcedBlock == iv.Hash {
+				sp.announcedBlock = nil
```

Comment:
I see that this hasn't changed from before, but what is the reasoning behind setting `announcedBlock` back to `nil` in this case?

Reply:
Yeah that'd looks like it'd remove the filter that just caught a duplicate, and `iv` could be sent again.

Reply:
Two reasons:

- The hash holds a reference into the inventory vector which means it can't be GCd until the reference is removed
- It's taking advantage of the fact that there are only two cases that a block announcement is made in this manner -- once in the tip checked (fast announce) case and once in the connect case.  So, it is only needed once.

Reply:
Makes sense, thanks for clarifying!